interop/test/ping.spec.ts

```diff
@@ -0,0 +1,164 @@
+        break
+      case 'ws':
+        options.transports = [webSockets()]
+        options.addresses = {
+          listen: isDialer ? [] : [`/ip4/${IP}/tcp/0/ws`]
+        }
+        break
+      case 'wss':
+        process.env.NODE_TLS_REJECT_UNAUTHORIZED = '0'
+        options.transports = [webSockets()]
+        options.addresses = {
+          listen: isDialer ? [] : [`/ip4/${IP}/tcp/0/wss`]
+        }
+        break
+      default:
+        throw new Error(`Unknown transport: ${TRANSPORT ?? '???'}`)
+    }
+
+    let skipSecureChannel = false
+    let skipMuxer = false
+    switch (TRANSPORT) {
+      case 'webtransport':
+      case 'webrtc':
+        skipSecureChannel = true
+        skipMuxer = true
+        break
+      default:
+        // Do nothing
+    }
+
+    if (!skipSecureChannel) {
+      switch (SECURE_CHANNEL) {
+        case 'noise':
+          options.connectionEncryption = [noise()]
+          break
+        case 'quic':
+          options.connectionEncryption = [noise()]
+          break
+        default:
+          throw new Error(`Unknown secure channel: ${SECURE_CHANNEL ?? ''}`)
+      }
+    } else {
+      // Libp2p requires at least one encryption module. Even if unused.
+      options.connectionEncryption = [noise()]
+    }
+
+    if (!skipMuxer) {
+      switch (MUXER) {
+        case 'mplex':
+          options.streamMuxers = [mplex()]
+          break
+        case 'yamux':
+          options.streamMuxers = [yamux()]
+          break
+        case 'quic':
+          break
+        default:
+          throw new Error(`Unknown muxer: ${MUXER ?? '???'}`)
+      }
+    }
+
+    const node = await createLibp2p(options)
+
+    try {
+      if (isDialer) {
+        let otherMa: string = (await redisProxy(['BLPOP', 'listenerAddr', timeoutSecs]).catch(err => { throw new Error(`Failed to wait for listener: ${err}`) }))[1]
+        // Hack until these are merged:
+        // - https://github.com/multiformats/js-multiaddr-to-uri/pull/120
+        otherMa = otherMa.replace('/tls/ws', '/wss')
+
+        console.error(`node ${node.peerId.toString()} pings: ${otherMa}`)
+        const handshakeStartInstant = Date.now()
+        await node.dial(multiaddr(otherMa))
+        const pingRTT = await node.services.ping.ping(multiaddr(otherMa))
+        const handshakePlusOneRTT = Date.now() - handshakeStartInstant
+        console.log(JSON.stringify({
+          handshakePlusOneRTTMillis: handshakePlusOneRTT,
+          pingRTTMilllis: pingRTT
+        }))
+      } else {
+        const multiaddrs = node.getMultiaddrs().map(ma => ma.toString()).filter(maString => !maString.includes('127.0.0.1'))
+        console.error('My multiaddrs are', multiaddrs)
+        // Send the listener addr over the proxy server so this works on both the Browser and Node
+        await redisProxy(['RPUSH', 'listenerAddr', multiaddrs[0]])
+        // Wait
+        await new Promise(resolve => setTimeout(resolve, 1000 * parseInt(timeoutSecs, 10)))
```

Comment:
This will always take the maximum amount of time, is there no way to wait for whatever it is we want to happen?  Otherwise these tests may spend a lot of time doing nothing.

Reply:
This is on purpose. The test runner will shutdown the listener when the dial completes. See https://github.com/libp2p/test-plans/tree/master/multidim-interop#listener